### PR TITLE
gateway-api: add mesh routes to RDS dependent configs

### PR DESCRIPTION
**Pull request by dddddai**
**Please provide a description of this PR:**
For http/tcp/tls mesh routes, we should be adding them(rather than the generated virtual services) to RDS dependent configs, otherwise the cache cannot get cleared correctly

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pkg/config/constants/constants.go [hunk 1]
```diff
@@ -120,8 +120,9 @@ const (
 	// Label to skip config comparison.
 	AlwaysPushLabel = "internal.istio.io/always-push"
 
-	// InternalParentName declares the original resource of an internally-generate config. This is used by ingress and the gateway-api.
-	InternalParentName     = "internal.istio.io/parent"
+	// InternalParentNames declares the original resources of an internally-generate config. This is used by k8s gateway-api.
+	// It is a comma separated list. For example, "HTTPRoute/foo.default,HTTPRoute/bar.default"
+	InternalParentNames    = "internal.istio.io/parents"
```
Comment (hzxuzhonghu) [inline pkg/config/constants/constants.go:125]: I think this is unrelated

File: pilot/pkg/networking/core/v1alpha3/route/route_cache.go [hunk 2]
```diff
@@ -96,7 +97,12 @@ func (r *Cache) DependentConfigs() []model.ConfigHash {
 		}.HashCode())
 	}
 	for _, vs := range r.VirtualServices {
-		configs = append(configs, model.ConfigKey{Kind: kind.VirtualService, Name: vs.Name, Namespace: vs.Namespace}.HashCode())
+		if model.UseGatewaySemantics(vs) {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/route/route_cache.go:100]: Do we miss this for Ingress?

File: pilot/pkg/networking/core/v1alpha3/route/route_cache.go [hunk 3]
```diff
@@ -96,7 +97,12 @@ func (r *Cache) DependentConfigs() []model.ConfigHash {
...
+			// add http/tcp/tls routes to dependent configs
+			configs = addInternalParents(configs, vs)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/route/route_cache.go:102]: This is not the right way for xds cache to be aware of gateway API. Except for gateway controller, other istio modules should be unaware of K8s Gateway/Ingress

File: pilot/pkg/networking/core/v1alpha3/route/route_cache.go (hunk 2 above)
Comment (dddddai) [reply pilot/pkg/networking/core/v1alpha3/route/route_cache.go:100]: AFAIK rds cache only applies to sidecars, so we don't need to care about ingress
But for gateway-api we are supporting **mesh** routes applying to sidecars

File: pilot/pkg/networking/core/v1alpha3/route/route_cache.go (hunk 3 above)
Comment (dddddai) [reply pilot/pkg/networking/core/v1alpha3/route/route_cache.go:102]: > Except for gateway controller, other istio modules should be unaware of K8s Gateway/Ingress

Ideally we should do this, but so far I don't see another way to clear cache on route updates

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route_cache.go:102]: I donot look into it, but this looks not good to me

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/route/route_cache.go:102]: @hzxuzhonghu I am not sure any other possibility. A Gateway-API VirtualService is synthetic and has dependencies outside itself. We could make it not say "GatewaySemantics" in the function name, but I am not sure how we can avoid this check.

Thinking about it more -- doesn't the same issue apply to `Ingress`?

Comment (dddddai) [reply pilot/pkg/networking/core/v1alpha3/route/route_cache.go:102]: > Thinking about it more -- doesn't the same issue apply to `Ingress`?

No, please see https://github.com/istio/istio/pull/40692#discussion_r957258617

**Review by howardjohn — APPROVED** [review]
